src/models.py

```diff
@@ -83,10 +84,44 @@ def fit(self, datagen, validation_datagen=None, meta_valid=None):
         self.callbacks.on_train_end()
         return self
 
-    def transform(self, datagen, validation_datagen=None):
+    def transform(self, datagen, validation_datagen=None, *args, **kwargs):
         outputs = self._transform(datagen, validation_datagen)
         for name, prediction in outputs.items():
-            outputs[name] = softmax(prediction, axis=1)
+            if isinstance(outputs[name], np.ndarray):
+                outputs[name] = softmax(prediction, axis=1)
+            else:
+                outputs[name] = [softmax(single_prediction, axis=1)[0] for single_prediction in prediction]
+        return outputs
+
+    def _transform(self, datagen, validation_datagen=None):
+        self.model.eval()
+
+        batch_gen, steps = datagen
+        outputs = {}
+        for batch_id, data in enumerate(batch_gen):
+            if isinstance(data, list):
+                X = data[0]
+            else:
+                X = data
+
+            if torch.cuda.is_available():
+                X = Variable(X, volatile=True).cuda()
+            else:
+                X = Variable(X, volatile=True)
+            outputs_batch = self.model(X)
+            if len(self.output_names) == 1:
+                outputs.setdefault(self.output_names[0], []).append(outputs_batch.data.cpu().numpy())
+            else:
+                for name, output in zip(self.output_names, outputs_batch):
+                    output_ = output.data.cpu().numpy()
+                    outputs.setdefault(name, []).append(output_)
+            if batch_id == steps:
+                break
+        self.model.train()
+        try:
```

Comment:
@apyskir I don't like the use of try catch here. I think we should have the private `self._trasnsform()` method return a list and if we want to we can vstack it in the public `self.transform` method. What do you think @apyskir @taraspiotr ?

Reply:
@apyskir I agree list of image predictions sounds best